### PR TITLE
Support no-std builds

### PR DESCRIPTION
**Pull request by str4d**

### CODE REVIEW

File: src/arithmetic/curves.rs [hunk 1]
```diff
@@ -62,6 +67,7 @@ pub trait CurveExt:
     ///     (g * x + &(h * r)).to_affine()
     /// }
     /// ```
+    #[cfg(feature = "alloc")]
     fn hash_to_curve<'a>(domain_prefix: &'a str) -> Box<dyn Fn(&[u8]) -> Self + 'a>;
```
Comment (ebfull) [inline src/arithmetic/curves.rs:71]: Technically it's not correct for a feature to enable a trait method for an exposed trait. Idea: make the whole `CurveExt` trait conditional on `std`. Only `halo2` right now needs `hash_to_curve` etc.

File: src/fields/fq.rs [hunk 2]
```diff
@@ -642,10 +649,23 @@ impl FieldExt for Fq {
     ]);
 
     fn sqrt_ratio(num: &Self, div: &Self) -> (Choice, Self) {
+        #[cfg(not(feature = "alloc"))]
+        {
+            let _ = (num, div);
+            unimplemented!()
+        }
```
Comment (ebfull) [inline src/fields/fq.rs:656]: Perhaps make an issue in this repo for resolving this.

**Review by ebfull — CHANGES_REQUESTED** [review]

File: src/arithmetic/curves.rs (hunk 1 above)
Comment (str4d) [reply src/arithmetic/curves.rs]: I moved `CurveExt` behind `alloc`, and `CurveAffine` behind `std`.

File: src/fields/fq.rs (hunk 2 above)
Comment (str4d) [reply src/fields/fq.rs]: Now that the entire `FieldExt` trait is behind the `std` feature flag, this is irrelevant.

File: src/arithmetic/fields.rs [hunk 3]
```diff
@@ -130,14 +138,66 @@ pub trait FieldExt: ff::PrimeField + From<bool> + Ord + Group<Scalar = Self> {
     }
 }
 
+/// Tonelli-Shank's square-root algorithm for `p mod 16 = 1`.
+///
+/// https://eprint.iacr.org/2012/685.pdf (page 12, algorithm 5)
+///
+/// `tm1d2` should be set to `(t - 1) // 2`, where `t = (modulus - 1) >> F::S`.
+#[cfg(not(feature = "std"))]
+pub(crate) fn sqrt_tonelli_shank<F: ff::PrimeField, S: AsRef<[u64]>>(
```
Comment (daira) [inline src/arithmetic/fields.rs:147]:
```suggestion
/// Tonelli-Shanks' square-root algorithm for `p mod 16 = 1`.
///
/// https://eprint.iacr.org/2012/685.pdf (page 12, algorithm 5)
///
/// `tm1d2` should be set to `(t - 1) // 2`, where `t = (modulus - 1) >> F::S`.
#[cfg(not(feature = "std"))]
pub(crate) fn sqrt_tonelli_shanks<F: ff::PrimeField, S: AsRef<[u64]>>(
```
[Shanks](https://en.wikipedia.org/wiki/Daniel_Shanks) is a name, not a possessive.

File: src/fields/fp.rs [hunk 4]
```diff
@@ -491,8 +506,22 @@ impl ff::Field for Fp {
...
-        let (is_square, res) = self.sqrt_alt();
-        CtOption::new(res, is_square)
+        #[cfg(feature = "std")]
+        {
+            let (is_square, res) = FP_TABLES.sqrt_alt(self);
+            CtOption::new(res, is_square)
+        }
+
+        #[cfg(not(feature = "std"))]
+        crate::arithmetic::sqrt_tonelli_shank(
```
Comment (daira) [inline src/fields/fp.rs:516]:
```suggestion
        crate::arithmetic::sqrt_tonelli_shanks(
```

File: src/fields/fq.rs [hunk 5]
```diff
@@ -491,8 +506,22 @@ impl ff::Field for Fq {
...
-        let (is_square, res) = self.sqrt_alt();
-        CtOption::new(res, is_square)
+        #[cfg(feature = "std")]
+        {
+            let (is_square, res) = FQ_TABLES.sqrt_alt(self);
+            CtOption::new(res, is_square)
+        }
+
+        #[cfg(not(feature = "std"))]
+        crate::arithmetic::sqrt_tonelli_shank(
```
Comment (daira) [inline src/fields/fq.rs:516]:
```suggestion
        crate::arithmetic::sqrt_tonelli_shanks(
```

File: src/arithmetic/fields.rs (hunk 3 above)
Comment (str4d) [reply src/arithmetic/fields.rs:147]: Whoops, I propagated this back from the original impl removed in https://github.com/zcash/halo2/pull/120.

**Review by nuttycom — APPROVED** [review]
utACK

**Pull request merged**